### PR TITLE
[REVIEW] upgrade CUB/Thrust to the latest commit

### PR DESCRIPTION
**Pull request by rongou**
This picks up two changes from CUB/Thrust:
* https://github.com/thrust/cub/pull/35: this gets rid of the patch command.
* https://github.com/thrust/thrust/pull/1255: this helps with https://github.com/NVIDIA/spark-rapids/issues/533.

@harrism @kkraus14

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by kkraus14** [comment]
I'm not comfortable with this change going in right before code freeze. @harrism?

**Comment by harrism** [comment]
@rongou This PR needs to target 0.16

File: cpp/src/copying/concatenate.cu [hunk 1]
```diff
@@ -70,14 +70,18 @@ auto create_device_views(std::vector<column_view> const& views, cudaStream_t str
...
-    device_views.cbegin(),
-    device_views.cend(),
-    std::next(offsets.begin()),
-    [](auto const& col) { return col.size(); },
-    thrust::plus<size_t>{});
+  thrust::transform(thrust::host,
+                    device_views.cbegin(),
+                    device_views.cend(),
+                    std::next(offsets.begin()),
+                    [](auto const& col) { return static_cast<size_t>(col.size()); });
```
Comment (jrhemstad) [inline cpp/src/copying/concatenate.cu:82]: The `size_t` above for the host_vector should be qualified as well. 

```suggestion
                    [](auto const& col) { return static_cast<std::size_t>(col.size()); });
```

File: cpp/src/copying/concatenate.cu [hunk 2]
```diff
@@ -70,14 +70,18 @@ auto create_device_views(std::vector<column_view> const& views, cudaStream_t str
   auto d_views = rmm::device_vector<column_device_view>{device_views};
 
   // Compute the partition offsets
+  // Note: Using 64-bit size_t so we can detect overflow of 32-bit size_type
+  // Note: Using separate transform and inclusive_scan because transform_inclusive_scan fails to
+  // compile with:
+  // error: no suitable constructor exists to convert from "size_t" to "cudf::column_device_view"
```
Comment (jrhemstad) [inline cpp/src/copying/concatenate.cu:76]: We should debug why we get this error first.

Comment (harrism) [reply cpp/src/copying/concatenate.cu:76]: Agree. We should not de-fuse these kernels if possible.

Comment (nvdbaranec) [reply cpp/src/copying/concatenate.cu:76]: As an aside, I was under the impression that nvcc had some facility to fuse multiple thrust calls like this into one effective kernel.  Is that not the case or is it just not as good as a proper hand-rolled one?

Comment (jrhemstad) [reply cpp/src/copying/concatenate.cu:76]: nvcc cannot fuse independent kernels into a single kernel.

Comment (rongou) [reply cpp/src/copying/concatenate.cu:76]: I think I have a fix for this in Thrust. Will send out a PR soon.

Comment (rongou) [reply cpp/src/copying/concatenate.cu:78]: https://github.com/thrust/thrust/issues/1260

**Comment by mike-wendt** [comment]
Doesn't look like we're needed for code review, removing ops

File: cpp/src/copying/concatenate.cu (hunk 2 above)
Comment (rongou) [reply cpp/src/copying/concatenate.cu]: Fixed.

**Comment by rongou** [comment]
rerun tests

**Comment by harrism** [comment]
> GIT_SHALLOW does not work with commit hashes

That's a good reason not to use commit hashes, IMO.

**Comment by harrism** [comment]
How much time does removing GIT_SHALLOW add to build time every time we run cmake? Because those of us who use rapids-compose run it *a lot*.

**Comment by rongou** [comment]
At least on my machine the time it takes to run `cmake` with an empty build directory seems to be comparable.

|branch-0.16|this PR|
|---|---|
|35.93|33.14|
|35.51|34.01|
|35.29|34.59|

With this PR it's actually slightly faster, probably because we no longer have to run the two patch commands.

**Pull request merged**